### PR TITLE
Add configurable resource requirements for Envoy sidecar in Webhook

### PR DESCRIPTION
**Pull request by miguelsorianod**
This PR adds configurable resource requirements for Envoy sidecar in Webhook.

Design:

The following annotations have been created that can be added into a ServiceDiscovery object.

```
        marin3r.3scale.net/resources.limits.cpu
        marin3r.3scale.net/resources.limits.memory
        marin3r.3scale.net/resources.requests.cpu
        marin3r.3scale.net/resources.requests.memory
```

The values have to follow the format of a Kubernetes `Quantity` https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#quantity-resource-core.

Because a `Quantity` cannot be specified as the empty string if a user specifies one of those annotation keys and it sets its value as the empty string `""` an error is returned at code level. This is one of the reasons the code implementation of the quantities parsing is surprisingly more complex than expected.

When an invalid `Quantity` is specified an error is returned. This is shown as an error in the `Deployment` object where the envoy sidecar is to be injected.

For example, if the user sets an invalid quantity value in the Deployment (kuard in our typical example) the following  condition is shown:
```
  - lastTransitionTime: "2020-10-30T19:30:16Z"
    lastUpdateTime: "2020-10-30T19:30:16Z"
    message: 'admission webhook "sidecar-injector.marin3r.3scale.net" denied the request:
      Error trying to load envoy container config from annotations: ''unable to parse
      quantity''s suffix'''
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
```

or also this one:

```
  - lastTransitionTime: "2020-10-30T19:44:36Z"
    lastUpdateTime: "2020-10-30T19:44:36Z"
    message: 'admission webhook "sidecar-injector.marin3r.3scale.net" denied the request:
      Error trying to load envoy container config from annotations: ''quantities must
      match the regular expression ''^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'''''
    reason: FailedCreate
```

Let me know your thoughts on this.

Pending:

- [x] Documentation
- [x] Testing

### CODE REVIEW

**Comment by miguelsorianod** [comment]
Addresses #47

File: pkg/webhooks/podv1mutator/sidecar.go [hunk 1]
```diff
@@ -125,9 +138,69 @@ func (esc *envoySidecarConfig) PopulateFromAnnotations(annotations map[string]st
...
+	}
+	esc.resources = resources
+
 	return nil
 }
 
+func lookupMarin3rAnnotation(key string, annotations map[string]string) (string, bool) {
+	value, ok := annotations[fmt.Sprintf("%s/%s", marin3rAnnotationsDomain, key)]
+	return value, ok
+}
```
Comment (roivaz) [inline pkg/webhooks/podv1mutator/sidecar.go:153]: Would be nice to use this small helper for all the code in this file WDYT?

**Comment by roivaz** [comment]
Looks pretty good to me. I agree that the validation of the quantities' syntax is important as it provides the user (through conditions in the Deployment or any other apps/v1 objects) with enough information to know where the error is and act on it.

File: pkg/webhooks/podv1mutator/sidecar.go (hunk 1 above)
Comment (miguelsorianod) [reply pkg/webhooks/podv1mutator/sidecar.go]: I agree.
I refactored the sidecar.go code to use it and I added the changes as a new commit

**Comment by miguelsorianod** [comment]
Added documentation and testing.

I think this is ready for a final review @roivaz

**Comment by roivaz** [comment]
@msoriano awesome work, thnx for the contrib! Could you rebase from master and push again? I have just enabled a couple of checks on the master branch and it needs the pipeline triggered again.

**Comment by miguelsorianod** [comment]
Rebased.

**Review by roivaz — APPROVED** [review]

**Pull request merged**